### PR TITLE
[mypyc] Precompute set literals for "in" ops against / iteration over set literals  

### PR DESCRIPTION
**Pull request by ichard26**
Towards https://github.com/mypyc/mypyc/issues/726. (There's a Python compatibility bug that needs to be fixed before the issue can be closed.)

For example, the set literals here are now precomputed as frozensets at module initialization.

    x in {1, 2.0, "3"}
    x not in {1, 2.0, "3"}

    for _ in {1, 2.0, "3"}:
        ...

Set literal items supported:

- Anything supported by `irbuild.constant_fold.constant_fold_expr()`
  - String and integer literals
  - Final references to int/str values
  - Certain int and str unary/binary ops that evaluate to a constant value
- `None`, `True`, and `False`
- Float, byte, and complex literals
- Tuple literals with only items listed above

**Results** (using gcc-9 on 64-bit Ubuntu)

Master @ 98cc165a657a316accb93f1ed57fdc128b086d9f

    running in_set
    ..........
    interpreted: 0.495790s (avg of 5 iterations; stdev 6.8%)
    compiled:    0.810029s (avg of 5 iterations; stdev 1.5%)

    compiled is 0.612x faster

    running set_literal_iteration
    .........................................................................................
    interpreted: 0.020255s (avg of 45 iterations; stdev 2.5%)
    compiled:    0.016336s (avg of 45 iterations; stdev 1.8%)

    compiled is 1.240x faster

This PR

    running in_set
    ..........
    interpreted: 0.502020s (avg of 5 iterations; stdev 1.1%)
    compiled:    0.390281s (avg of 5 iterations; stdev 6.2%)

    compiled is 1.286x faster

    running set_literal_iteration
    ..............................................................................................
    interpreted: 0.019917s (avg of 47 iterations; stdev 2.2%)
    compiled:    0.007134s (avg of 47 iterations; stdev 2.6%)

    compiled is 2.792x faster

Benchmarks can be found here: mypyc/mypyc-benchmarks#32

### CODE REVIEW

**Comment by ichard26** [comment]
FYI, I wrote the PR description to be used as the commit message when this is squash merged into main. Please copy and paste it and save yourself some work! :)

**Comment by ichard26** [comment]
I'll push a functional work around to the mypyc bug that's breaking the build sometime later, but for those wondering "do people really loop over set literals often?" [here's a GitHub search query for you](https://github.com/search?type=code&q=language%3Apython%20%2Ffor%20%5Ba-zA-Z%5D%2B%20in%20%5C%7B.%2B%5C%7D%2F) (there's even an example in Black's source code :slightly_smiling_face:).

File: mypyc/lib-rt/misc_ops.c [hunk 1]
```diff
@@ -635,6 +636,22 @@ int CPyStatics_Initialize(PyObject **statics,
...
+        while (num-- > 0) {
+            int num_items = *frozensets++;
+            PyObject *obj = PyFrozenSet_New(NULL);
+            if (obj == NULL) {
+                return -1;
+            }
+            for (int i = 0; i < num_items; i++) {
+                PyObject *item = statics[*frozensets++];
+                Py_INCREF(item);
+                PySet_Add(obj, item);
```
Comment (JukkaL) [inline mypyc/lib-rt/misc_ops.c:650]: Check the return value of `PySet_Add`.

Comment (ichard26) [reply mypyc/lib-rt/misc_ops.c]: Whoops I totally forgot about this. Good catch! Addressed in https://github.com/python/mypy/pull/14409/commits/9cdfc98b25058f0bd4c8e45c44aeea97dd46a04e.

**Review by JukkaL — APPROVED** [review]

**Pull request merged**